lib/auth/keystore/keystore_test.go

```diff
@@ -0,0 +1,288 @@
+fPTgihJAeKdWbBmRMjIDe8hkz/oxR6JE2Ap+4G+KZtwVON4b+ucCYTQS+1CQp2Xc
+RPAMyjbzPhWQpfJnIxLcqGmvXxosABvs/b2CWaPqfCQhZIWpLeKW
+-----END RSA PRIVATE KEY-----
+`)
+	testRawPublicKey = []byte("ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQCqIPatEnmSruE/nl44Iz0O12RY8wjw8EqDFoCJD0n1gXdo26v3xvyQAJxNDyLFKQewtxxWpF238KdvuO0FKvDEn0E3T19kSLIK/LuXDE0YExARqi/1RTvbae+5Ypv4SQfNxVX2J72LzAzzzGnPNg7x51rwVmMfFpqh7aWSseKI4VBG5smodWFb5I0VA5Xo6xURNNmWDmuZaEDmsqIHobRB4sfKxIwltssw5evVVu7tGqiGarQAXoR0yCLHc4nPeov1gMpA8DOGPtWI/NPTs+//2+Hl+NdoTmJOE9Piffe5jU3Z8kCfOxxm9WanHG5I6rHBYGqRHMgl7PW+/cX7nEMv")
+	testRawCert      = []byte(`-----BEGIN CERTIFICATE-----
+MIIDeTCCAmGgAwIBAgIRALmlBQhTQQiGIS/P0PwF97wwDQYJKoZIhvcNAQELBQAw
+VjEQMA4GA1UEChMHc2VydmVyMTEQMA4GA1UEAxMHc2VydmVyMTEwMC4GA1UEBRMn
+MjQ2NzY0MDEwMjczNTA2ODc3NjY1MDEyMTc3Mzg5MTkyODY5ODIwMB4XDTIxMDcx
+NDE5MDY1MloXDTMxMDcxMjE5MDY1MlowVjEQMA4GA1UEChMHc2VydmVyMTEQMA4G
+A1UEAxMHc2VydmVyMTEwMC4GA1UEBRMnMjQ2NzY0MDEwMjczNTA2ODc3NjY1MDEy
+MTc3Mzg5MTkyODY5ODIwMIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEA
+qiD2rRJ5kq7hP55eOCM9DtdkWPMI8PBKgxaAiQ9J9YF3aNur98b8kACcTQ8ixSkH
+sLccVqRdt/Cnb7jtBSrwxJ9BN09fZEiyCvy7lwxNGBMQEaov9UU722nvuWKb+EkH
+zcVV9ie9i8wM88xpzzYO8eda8FZjHxaaoe2lkrHiiOFQRubJqHVhW+SNFQOV6OsV
+ETTZlg5rmWhA5rKiB6G0QeLHysSMJbbLMOXr1Vbu7Rqohmq0AF6EdMgix3OJz3qL
+9YDKQPAzhj7ViPzT07Pv/9vh5fjXaE5iThPT4n33uY1N2fJAnzscZvVmpxxuSOqx
+wWBqkRzIJez1vv3F+5xDLwIDAQABo0IwQDAOBgNVHQ8BAf8EBAMCAqQwDwYDVR0T
+AQH/BAUwAwEB/zAdBgNVHQ4EFgQUAJprFmwUDaYguqQJxHLBC35BeQ0wDQYJKoZI
+hvcNAQELBQADggEBAG3k42nHnJvsf3M4EPBqMFqLHJOcwt5jpfPrjLmTCAnjialq
+v0qp/JAwC3SgrDFQMNNcWTi+H+E9FqYVavysZbBd0/cFlYH9SWe9CJi5CyfbsLGD
+PX8hBRDZmmhTXMMHzyHqolVAFCf5nNQyQVeQGt3fBDh++WNjmkS+886lhag/a2hh
+nskCVzvig1/6Exp06k2mMlphC25/bNB3SDeQj+dIJdej6btZs2goQZ/5Sx/iwB5c
+xrmzqBs9YMU//QIN5ZFE+7opw5v6mbeGCCk3woH46VmVwO6mHCfLha4K/K92MMdg
+JhuTMEqUaAOZBoQLn+txjl3nu9WwTThJzlY0L4w=
+-----END CERTIFICATE-----
+`)
+	testPKCS11Key = []byte(`pkcs11:{"host_id": "server2", "key_id": "00000000-0000-0000-0000-000000000000"}`)
+
+	testRawSSHKeyPair = &types.SSHKeyPair{
+		PublicKey:      testRawPublicKey,
+		PrivateKey:     testRawPrivateKey,
+		PrivateKeyType: types.PrivateKeyType_RAW,
+	}
+	testRawTLSKeyPair = &types.TLSKeyPair{
+		Cert:    testRawCert,
+		Key:     testRawPrivateKey,
+		KeyType: types.PrivateKeyType_RAW,
+	}
+	testRawJWTKeyPair = &types.JWTKeyPair{
+		PublicKey:      testRawPublicKey,
+		PrivateKey:     testRawPrivateKey,
+		PrivateKeyType: types.PrivateKeyType_RAW,
+	}
+
+	testPKCS11SSHKeyPair = &types.SSHKeyPair{
+		PublicKey:      testRawPublicKey,
+		PrivateKey:     testPKCS11Key,
+		PrivateKeyType: types.PrivateKeyType_PKCS11,
+	}
+	testPKCS11TLSKeyPair = &types.TLSKeyPair{
+		Cert:    testRawCert,
+		Key:     testPKCS11Key,
+		KeyType: types.PrivateKeyType_PKCS11,
+	}
+	testPKCS11JWTKeyPair = &types.JWTKeyPair{
+		PublicKey:      testRawPublicKey,
+		PrivateKey:     testPKCS11Key,
+		PrivateKeyType: types.PrivateKeyType_PKCS11,
+	}
+)
+
+func TestKeyStore(t *testing.T) {
```

Comment:
Is this intended to be an interface compliance test for multiple implementations of `KeyStore`?

Reply:
This test seems to be exercising a _lot_ of functionality in one place. In a perfect world I would suggest trying to split it up somehow, but I don't think it would be worth blocking the PR for it.

Reply:
Yep, you can see the added HSM testcases in the following PR https://github.com/gravitational/teleport/pull/7614